src/software/world/timestamped_possession_state.h

```diff
@@ -0,0 +1,106 @@
+#pragma once
+
+#include <map>
+#include <optional>
+
+#include "software/time/timestamp.h"
+#include "software/world/robot_state.h"
+#include "software/world/team_types.h"
+
+/**
+ * A light structure for a robot ID with TeamSide
+ */
+struct RobotIdWithTeamSide
+{
+    unsigned int id;
+    TeamSide team_side;
+
+    bool operator==(const RobotIdWithTeamSide &other) const
+    {
+        return id == other.id && team_side == other.team_side;
+    }
+
+    bool operator<(const RobotIdWithTeamSide &other) const
+    {
+        return id < other.id || team_side < other.team_side;
+    }
+};
+
+class TimestampedPossessionState
+{
+   public:
+    /**
+     * Creates a new possession state that tracks no robots' possession and has timestamp
+     * of 0
+     */
+    TimestampedPossessionState();
+
+    /**
+     * Updates the possession state
+     *
+     * @throws std::invalid_argument if update_time is older than timestamp()
+     *
+     * @param possessions The robots that have possession of the ball at update_time
+     * @param update_time The timestamp to update possession state
+     */
+    void updateState(std::vector<RobotIdWithTeamSide> possessions, Timestamp update_time);
+
+    /**
+     * Returns the timestamp of when the possession state was last updated
+     *
+     * @return the timestamp
+     */
+    Timestamp timestamp() const;
+
+    /**
+     * Returns the robots that currently have possession of the ball. Usually, this will
+     * be only one robot, but multiple robots could be fighting over the ball. If there
+     * isn't enough data or if no robots have possession, then returns an empty vector
+     *
+     * @return The vector of RobotIdWithTeamSide with possession of the ball
+     */
+    const std::vector<RobotIdWithTeamSide> &getRobotsWithPossession() const;
```

Comment:
I'm not a huge fan of the `RobotIdWithTeamSide` in the public interface. I can see why it may be useful internally to avoid code duplication (like I ended up doing with the simulation stuff for yellow/blue teams), but I think in most cases it's much easier to just have different functions for each team in the public interface?

I think `getFriendlyRobotWithPossession` would be more useable than `getRobotsWithPossession` and then checking for a matching ID and TeamSide. I think the former makes the consumer code cleaner because there's fewer conditions to check, and I can't think of any reasons consumers would need the info for _all_ robots with possession atm.

Reply:
I think users would end up calling both getFriendlyWithPossession and getEnemywithPossesion anyway, since most users will want all of that information, so I don't see how getFriendlyRobotWithPossession is more usable. Perhaps this is larger discussion of our _with_ data types, e.g. RobotStateWithId, which is similar idea as a public interface

Reply:
Hmm, this a tricky one. I imagine that in most cases we'll really care about one of two cases:
1) Does the friendly/enemy team have possession
2) Which robot has possession

I think that if we really care which robot actually has possession, then we probably want a `Robot` class back? Rather then having to go lookup the robot in whichever team it's from?

Reply:
My point is that even if a user needs to call both `getFriendlyWithPossession` and `getEnemywithPossesion`, I think it's still easier to use because "checking for" what team the robot is a part of is already implicitly done by what function is called.

So the user could do
```
auto friendly_opt = getFriendlyWithPossession();
auto enemy_opt = getEnemyWithPossession();
```
rather than
```
auto robots = getRobotsWithPossession();
auto friendly_iter = std::find(...)
if(friendly_iter != robots.end()) {
   auto friendly = *friendly_iter;
}
auto enemy_iter = std::find(...)
if(enemy_iter != robots.end()) {
   auto enemy = *enemy_iter;
}
```
which I think is much easier to use, because it at least saves a `find()`

Because of the assumption / invariant that no more than one robot per team can have possession at a time, returning an (optional) single robot per team removes any assumptions about the size of the returned vector or any code that has to handle a vector of any size.

Reply:
why should it be a single robot per team? What if two enemy robots are fight over the ball or are somehow sharing possession?

Reply:
I'll change it to per team but I think we should return a vector

Reply:
@garethellis0 use your brain. Think of an interface

Reply:
IMO possession is exclusive. There should only ever be one robot with possession. If two robots are fighting over a ball then by definition neither of them have possession. I think we should have:
```
std::optional<Robot> World::getRobotWithPossession();

std::optional<Robot> FriendlyTeam::getRobotWithPossession();

std::optional<Robot> EnemyTeam::getRobotWithPossession();
```

I get the desire to have concepts like shared possession, but IMO they introduce too much semantic and syntactic complexity. At that point we'd probably want to add `exclusivePossesion` functions as well, which really just seems like more work then it's worth.